git-versions.yml

```diff
@@ -1,17 +1,17 @@
 # This file lists all software components that are part of a Mender
-# release with their internal versions. The keys represent images
-# rather than "services", but we keep it for the sake of parsing it
-# the same way as other-components and docker-compose files.
+# release with their git versions. The keys represent images rather
+# than "services" (or containers), but we keep it for parsing it the
+# same way as other-components and docker-compose files.
 services:
 
     #
-    # backend open source services
+    # backend open source images
     #
     deployments:
         git-version: master
     gui:
         git-version: master
-    mender-api-gateway-docker:
+    api-gateway:
```

Comment:
I think this should be the container name (`mender-api-gateway`), not the image name, so that the format of this file matches what is in docker-compose.yml. The previous `--version-of` command worked for all three names, so I think it should keep working like that.

In relation to that I just thought of a completely different fix which is way simpler than the current approach: Instead of loading both YML files and keeping separate entries, just load *either* `docker-compose.yml` or `git-versions.yml` depending on which one is desired, and return direct results from there. No need to have separate `git-version` entries, just make sure the files are the same format. With this we could remove quite a lot of special case logic. What do you think?

Reply:
Thanks. I think is good that we are iterating the design so that we might end up with a simpler one :)

> I think this should be the container name (`mender-api-gateway`), not the image name, so that the format of this file matches what is in docker-compose.yml.

In the first iteration I was using repos names here, but it was very hard to merge the info while parsing the rest of the yaml files (see below wrt re: parsing them separately).

The problem with your suggestion is that the container names are not unique. In the old code, for each container we extract the prefix/image/version parameters and then use the "image" as the unique key for our data. We could take some shortcuts, for example if we force enterprise and os to have always the same versions, then we could unify the container names.

> The previous `--version-of` command worked for all three names, so I think it should keep working like that.

Yes, I would like to keep it too. It is clear that I added the restriction to compensate the shortcoming of this design. If we manage to change it, we can take out the restriction.

> 
> In relation to that I just thought of a completely different fix which is way simpler than the current approach: Instead of loading both YML files and keeping separate entries, just load _either_ `docker-compose.yml` or `git-versions.yml` depending on which one is desired, and return direct results from there. No need to have separate `git-version` entries, just make sure the files are the same format. With this we could remove quite a lot of special case logic. What do you think?

And this is something that in principle I like too, and maybe you have a good suggestion for the following problem.

The issue here is that not all components exist in the git-versions.yml file. The reasoning by me was to not duplicate version information: we would be using git-versions.yml only for the images that use different tags for docker/git (namely backend images) while still taking the version from the image specification for the rest (mender client, mender-cli, mender-artifact).

Reply:
> The problem with your suggestion is that the container names are not unique. In the old code, for each container we extract the prefix/image/version parameters and then use the "image" as the unique key for our data.

Can't we still do that? Use the exact same format, as in:

```
docker-compose.yml:
    mender-api-gateway:
        image: mendersoftware/api-gateway:mender-master

git-versions.yml:
    mender-api-gateway:
        image: mendersoftware/api-gateway:master
```

Then `image` can be used as the unique identifier as before. There will be one piece of duplicated information, and that is the container name, but I think that price is worth paying for keeping the parsing logic the same everywhere, as it's already complicated enough. Besides, the container name already exists as an "extra" component in the `other-components.yml` file, so it's not really new.

> The issue here is that not all components exist in the git-versions.yml file. The reasoning by me was to not duplicate version information: we would be using git-versions.yml only for the images that use different tags for docker/git (namely backend images) while still taking the version from the image specification for the rest (mender client, mender-cli, mender-artifact).

I agree that we should not duplicate version information. But I don't think we need to, we can just maintain a map of docker-only and git-only files:

```
DOCKER_ONLY_YML = ["docker-compose.yml", "docker-compose.enterprise.yml"]
GIT_ONLY_YML = ["git-versions.yml"]
```

and the rest of the YML files we always parse. This would ensure that the non-docker tools are always loaded, because they are in other files (`other-components.yml` and `docker-compose.client.yml`), and the docker-versions and git-versions are loaded only when they are needed.

Does this solve the problem, you think?

Reply:
Yes, that will overall work. The only side effect is that we need a separate `git-versions-enterprise.yml` to not override containers keys, but we can live with that.